File: backend/src/utils/embedding/adapters/openai.ts
import type { EmbeddingAdapter, EmbeddingProviderRecord } from '../types'
import { DEFAULT_MODELS, buildBearerHeaders, parseOpenAIResponse } from './base'

export class OpenAIAdapter implements EmbeddingAdapter {
  getEndpointPath(_provider: EmbeddingProviderRecord): string {
    return '/embeddings'
  }

  buildHeaders(provider: EmbeddingProviderRecord): Record<string, string> {
    return buildBearerHeaders(provider.api_key)
  }

  buildBody(provider: EmbeddingProviderRecord, input: string): Record<string, unknown> {
    const body: Record<string, unknown> = { input }
    const model = provider.model || DEFAULT_MODELS['openai']
    if (model) body['model'] = model
    if (provider.dimensions !== null) body['dimensions'] = provider.dimensions
    return body
  }

  parseResponse(data: unknown): number[] {
    return parseOpenAIResponse(data)
  }
}

File: gateway-engine/src/adapters/embedding/openai.rs
//! OpenAI embedding adapter — standard `/embeddings` endpoint, Bearer auth.

use serde_json::Value;

use crate::adapters::bearer_headers;
use super::{parse_openai_response, standard_body, EmbeddingAdapter};
use crate::agents::embedding::client::EmbeddingProviderConfig;

pub struct OpenAiEmbeddingAdapter;

impl EmbeddingAdapter for OpenAiEmbeddingAdapter {
    fn vendor(&self) -> &'static str {
        "openai"
    }

    fn endpoint_path(&self, _p: &EmbeddingProviderConfig) -> String {
        "/embeddings".to_string()
    }

    fn build_headers(&self, p: &EmbeddingProviderConfig) -> Vec<(String, String)> {
        bearer_headers(p.api_key.as_deref())
    }

    fn build_body(&self, p: &EmbeddingProviderConfig, input: &str) -> Value {
        standard_body(p, input)
    }

    fn parse_response(&self, data: &Value) -> Result<Vec<f32>, String> {
        parse_openai_response(data)
    }
}
